### ISSUE TITLE
What would you incompatibly break?

### ISSUE THREAD

**Issue by kstrauser** [schematics/schematics#324]
We've learned a lot along the way. Knowing what you know now, if you were implementing Schematics from scratch, what would you do differently? I'm *not* suggest we do this - I was just discussing this with @jvantuyl and it was an interesting exploration.

I'll lead:

- I'd toss Python 2.6 compatibility. I don't think there's much demand for it anymore and it's officially retired.
- Implement @bintoro's PR #266 metadata patch or something very similar.

Your turn. What would you personally do differently if this were a brand new project?

**Comment by jvantuyl** [comment]
First and foremost, I don't want to give the impression that schematics isn't a wonderful piece of engineering.  I use it every day and it does what it was designed to do... quite well.

That said, Python 3 has brought with it quite a few changes.  Some of the data I've encountered legitimately has cycles.  Other data has pretty unusual requirements around correctness (i.e. a dict with a key that could be an int, a string, or another dict with a certain schema).  Finally, there are some dark corners of the codebase where we "do the right thing" that can sometimes manifest as "behavior" (Random references:  #53 + #54 + #76 [semi-intentional sloppy coercion across all of our types], #167 + #184 + #185 + #186 [polymorphic data structures are kind of a pain], #194 [bytes makes this complicated again], #210 [no easy way to validate a whole model without piggy-backing on a field or type validator], #266 + #316 + #318 [issues right at the heart of composition / strictness]).

There's a lot to be done that can adapt to the above issues.  Off the top of my head:

* Make coercion more controllable (i.e. right now, our models have optional strict interpretation but our fields are always loose)
* Make StringFields encoding-smart (i.e. encoding=None [which passes through a string] versus encoding=utf_8/ascii/utf_16/shift_jis/anything_in_codecs/rot13_lol).
* Make ByteFields for when we actually want untouched byte strings.
* Perhaps a relation API to sew together different datasets that have references.
* Find a way to solidly address the concerns of polymorphic data (this is easier when encoding than when decoding)
* Global validators on models.
* Less magic.  We're have a lot of complexity for what we get out of things like meta-classes, I think.
* Perhaps trade our transform loops for some recursion?  I can't help but think that the polymorphism stuff would be easier if we could delegate to the models for some of the composition.

There's probably plenty more, but the overall difference I'm shooting for is:

* take a step back and see if we can refactor the core to better tackle the things it makes difficult
* fit better into Python 3 (the bytes / strings thing is new, very touchy, and encoding is our concern)
* make schematics easier to control when we're strict and when we're sloppy
* make schematics easier to extend (this is basically black-magic beyond simple fields)
* generate more / better ways to make comprehensive validations
* easier to make sophisticated compositions of different schematics in a single data-structure
* easier to know what's going on under the hood--especially with ModelTypes and the like

Anyways, that's just my 0.02¢.  I'd imagine that this is going to take some concrete proposals before we get anywhere.  I'm curious what others have to say.

**Comment by lkraider** [comment]
I agree on the correctness point, I even started to model the algorithm in TLA+ formal language, to try to pin down the core (see my doc-math branch here https://github.com/nKey/schematics/tree/doc-math/docs/math ). This would allow model proofs on the algorithm which the implementation could follow.

On that point, I have my own branch which has some guarantees over integrity that are not merged into master, and also not updated to the latest schematics here, but which I use in production. For me they are required for the core to work correctly. See my (old) pull request here: https://github.com/schematics/schematics/pull/124

**Comment by bintoro** [comment]
This is an important discussion to have.

I'm going to link to #263 here. The issue was left in limbo because initially it seemed that a breaking change would be necessary.

Eventually, I came up with [a third approach](https://github.com/schematics/schematics/issues/263#issuecomment-69474058) that would largely preserve compatibility with existing applications. But if we're starting to plan Schematics 2.0, it means all the options are on the table as far as this feature goes.

**Comment by bintoro** [comment]
I'm about to start merging backward-incompatible features into the main branches, and therefore Schematics v1 will move on to live on a separate legacy branch. It can still be maintained, but after v1.2.0 the main focus will be on moving things forward by dealing with the current backlog of PRs and issues, most of which are more or less backward-incompatible.

Before I let the branches diverge, I'd like to get all the non-breaking, non-controversial PRs out of the way. I've already merged a few contributions, but then there are these:

* #349 and #350 I consider to be bugfixes unless someone can explain otherwise. These I'm planning to incorporate in v1, so a review would be appreciated.

* #355 introduces functionality that should exist according to documentation. It will only break things in what I believe to be a rare case, so I'm thinking it could go in v1 as well.

* #351 (912afcc02c5506290d8fe5d818207c141cc6d1e3) changes the primitive value from `None` to `{}` when a model collapses into nothing when all its fields are empty (`None` or `{}` or `[]`) and `serialize_when_none` is `False`. This doesn't break compatibility too badly either, but since the handling of empty values is incomplete anyway, maybe this should appear in v2 along with the Undefined feature.

**Comment by kstrauser** [comment]
Ooh, 👍🏻x10 for #351 and #355!

Something @jvantuyl brought up: is there any real reason why we have the import and export loops instead of recursion? Do they buy us anything?

**Comment by lkraider** [comment]
Loops don't tend to explode the stack as recursion does.

On Monday, 9 November 2015, Kirk Strauser <notifications@github.com> wrote:

> Ooh, 👍🏻x10 for #351 <https://github.com/schematics/schematics/pull/351>
> and #355 <https://github.com/schematics/schematics/pull/355>!
>
> Something @jvantuyl <https://github.com/jvantuyl> brought up: is there
> any real reason why we have the import and export loops instead of
> recursion? Do they buy us anything?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/schematics/schematics/issues/324#issuecomment-155234456>
> .
>


-- 

--
Paul Eipper

**Comment by bintoro** [comment]
> Ooh, 👍🏻x10 for #351 and #355!

I actually withdrew #355 in response to @lkraider's [objection](https://github.com/schematics/schematics/pull/355#issuecomment-154555985), which totally makes sense. It's probably better to bundle value-munging with the conversion step rather than with validation.

> Something @jvantuyl brought up: is there any real reason why we have the import and export loops instead of recursion? Do they buy us anything?

I don't understand the question... The loops provide the recursion, no?

**Comment by j2labs** [comment]
The loops recurse into compound models but loop across fields.

Hopefully someone's data structure isnt so deep that they blow the stack
there... ;)

On Wednesday, November 11, 2015, Kalle Tuure <notifications@github.com>
wrote:

> Ooh, 👍🏻x10 for #351 <https://github.com/schematics/schematics/pull/351>
> and #355 <https://github.com/schematics/schematics/pull/355>!
>
> I actually withdrew #355
> <https://github.com/schematics/schematics/pull/355> in response to
> @lkraider <https://github.com/lkraider>'s objection
> <https://github.com/schematics/schematics/pull/355#issuecomment-154555985>,
> which totally makes sense. It's probably better to bundle value-munging
> with the conversion step rather than with validation.
>
> Something @jvantuyl <https://github.com/jvantuyl> brought up: is there
> any real reason why we have the import and export loops instead of
> recursion? Do they buy us anything?
>
> I don't understand the question... The loops provide the recursion, no?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/schematics/schematics/issues/324#issuecomment-155785338>
> .
>


-- 
jms dnns
me <http://jmsdnns.com> // facebook <https://www.facebook.com/jmsdnns> //
github <https://github.com/jmsdnns/> // twitter
<https://twitter.com/jmsdnns>

**Comment by kstrauser** [comment]
I can't imagine and structure where that's actually a concern. Even the standard library throws its hands in the air after a certain depth:

```python
>>> a = {}
>>> for i in range(1000): a = {'': a}
>>> import pprint
>>> pprint.pprint(a)
...
RuntimeError: maximum recursion depth exceeded
```

Remember, the stack would only care about maximum *depth*. Make a model with 10,000 fields and a recursive approach wouldn't blink an eye.

My point about ditching the export loop in favor of recursion is that I've always approached the loop with a touch of fear. It also seemed like that should instead be a `.export` method like:

```python
class StringType(object):
    def export(self):
        return repr(self)

class ListType(object):
    def export(self):
        return [child.export() for child in self.children]
```

...and a corresponding `.import` method. It always "felt" odd to me that an external loop has to introspect each object to do the right thing.

I hadn't seen that about #355. I kind of like that pattern, but the logic makes sense.

**Comment by bintoro** [comment]
> My point about ditching the export loop in favor of recursion is that I've always approached the loop with a touch of fear. It also seemed like that should instead be a `.export` method

But isn't this almost exactly how it already works? It's just that `export` is currently named either `to_primitive` or `export_loop` depending on whether it's a basic or compound type. The two methods have diverged sometime in the past simply because their usage got different. I want to merge them again, which becomes possible once #266 is finished. See also #337.

> It always "felt" odd to me that an external loop has to introspect each object to do the right thing.

If you're talking about this:

```python
if hasattr(field, 'export_loop'):
  ...
else:
  ...
```

then I agree this is totally crazy and precisely what got me started on #266.

If you're talking about why there is an external `export_loop` in the `transforms` module, then that's because the main loops are general import/export processors that can act on arbitrary mappings as well as actual `Model` instances.

For example, `validate()` invokes `import_loop` to run a model's validators on a dictionary, and the response is a modified dictionary.

An even better example is the `Model.import_data()` method, which might ask `import_loop` to perform import conversion on a dictionary consisting of just one item.

Were it placed in the `Model` class, `import_loop` would essentially be a static method anyway. So I think the current arrangement seems pretty reasonable.

**Comment by j2labs** [comment]
I am the original author of Schematics, so I can say clearly why I made
those decisions.

By not putting those methods inside classes, we don't need to have
instances of classes. One could use a static method, but I didn't want to
encourage any more OO than necessary for the metaprogramming to work.

Recursion simply isn't necessary for exporting fields. It doesn't actually
make sense to me how this would work since fields are stored in a list that
is easily iterated across.

I didn't write the code where we see `if hasattr(field, 'export_loop')`, so
I can't explain that, but I would prefer a cleaner approach here too. I
think maybe the ModelType fields could handle this themselves via calls to
`to_primitive` or `to_native`.

The reason there is no blank `export` or `import` is because I want users
to be aware of whether or not they're converting to python types or to
types that are safe to put in json conversion tools.

To be clear, the transforms module exists because the initial
implementation of the transforms were not attached to the models at all. We
put those functions  in there (well, Jokull and co did) to get users the
functionality with fewer imports.


On Wed, Nov 11, 2015 at 1:07 PM, Kalle Tuure <notifications@github.com>
wrote:

> My point about ditching the export loop in favor of recursion is that I've
> always approached the loop with a touch of fear. It also seemed like that
> should instead be a .export method
>
> But isn't this almost exactly how it already works? It's just that export
> is currently named either to_primitive or export_loop depending on
> whether it's a basic or compound type. The two methods have diverged
> sometime in the past simply because their usage got different. I want to
> merge them again, which becomes possible once #266
> <https://github.com/schematics/schematics/pull/266> is finished. See also
> #337 <https://github.com/schematics/schematics/issues/337>.
>
> It always "felt" odd to me that an external loop has to introspect each
> object to do the right thing.
>
> If you're talking about this:
>
> if hasattr(field, 'export_loop'):
>   ...else:
>   ...
>
> then I agree this is totally crazy and precisely what got me started on
> #266 <https://github.com/schematics/schematics/pull/266>.
>
> If you're talking about why there is an external export_loop in the
> transforms module, then that's because the main loops are general
> import/export processors that can act on arbitrary mappings as well as
> actual Model instances.
>
> For example, validate() invokes import_loop to run a model's validators
> on a dictionary, and the response is a modified dictionary.
>
> An even better example is the Model.import_data() method, which might ask
> import_loop to perform import conversion on a dictionary consisting of
> just one item.
>
> Were it placed in the Model class, import_loop would essentially be a
> static method anyway. So I think the current arrangement seems pretty
> reasonable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/schematics/schematics/issues/324#issuecomment-155863704>
> .
>



-- 
jms dnns
me <http://jmsdnns.com> // facebook <https://www.facebook.com/jmsdnns> //
github <https://github.com/jmsdnns/> // twitter
<https://twitter.com/jmsdnns>

**Comment by bintoro** [comment]
This touches on something I thought about a few days ago... What if `to_primitive()` and `to_native()` were defined only in `BaseType` and `MultiType`, and they would defer to the appropriate private method such as `_to_primitive()` on the intended field?

The original motivation was to make importing and exporting extensible. It would be easy to hook up new functionality to fields if the method calls always went through the base class.

But this could also be exploited to hide the `export_loop`/`to_primitive` dichotomy:

* Calling `StringType.to_primitive()` would hit `BaseType.to_primitive()`, which would then call the actual target `StringType._to_primitive()`.

* Calling `ListType.to_primitive()` would invoke `MultiType.to_primitive()`, which would call `ListType.export_loop()`.

**Comment by j2labs** [comment]
Yes! I agree with this!

On Wednesday, November 11, 2015, Kalle Tuure <notifications@github.com>
wrote:

> This touches on something I thought about a few days ago... What if
> to_primitive() and to_native() were defined only in BaseType and MultiType,
> and they would defer to the appropriate private method such as
> _to_primitive() on the intended field?
>
> The original motivation was to make importing and exporting extensible. It
> would be easy to hook up new functionality to fields if the method calls
> always went through the base class.
>
> But this could also be exploited to hide the export_loop/to_primitive
> dichotomy:
>
>    -
>
>    Calling StringType.to_primitive() would hit BaseType.to_primitive(),
>    which would then call the actual target StringType._to_primitive().
>    -
>
>    Calling ListType.to_primitive() would invoke MultiType.to_primitive(),
>    which would call ListType.export_loop().
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/schematics/schematics/issues/324#issuecomment-155924477>
> .
>


-- 
jms dnns
me <http://jmsdnns.com> // facebook <https://www.facebook.com/jmsdnns> //
github <https://github.com/jmsdnns/> // twitter
<https://twitter.com/jmsdnns>

**Comment by kstrauser** [comment]
@j2labs Well, I wouldn't recurse across an model's own fields - I'd iterate across them. Recursion would come into play if one of its fields was a container, in which case exporting that field would recurse into it.

@bintoro That's very similar to what I had in mind.

**Comment by j2labs** [comment]
> @j2labs <https://github.com/j2labs> Well, I wouldn't recurse across an
> model's own fields - I'd iterate across them. Recursion would come into
> play if one of its fields was a container, in which case exporting that
> field would recurse into it.
>
This is already what happens. No?

>

-- 
jms dnns
me <http://jmsdnns.com> // facebook <https://www.facebook.com/jmsdnns> //
github <https://github.com/jmsdnns/> // twitter
<https://twitter.com/jmsdnns>

**Comment by bintoro** [comment]
@jmsdnns, @kstrauser, any thoughts on the proposed patch #368?

To summarize, it's a variant of [this idea](https://github.com/schematics/schematics/issues/324#issuecomment-155924477) that already got tentative support, but instead of using `to_primitive` and `to_native` as the universal interface I landed on `export` and `convert`.

The reason is that the *format* (primitive/native) is just a detail that is subordinate to the *process* (import/export). Compound types don't care about the format at all. It's also more flexible this way, 'cause now the formats aren't hardcoded in the interface. A custom type can implement a third format if it wants.

This is what `ListType.export` (formerly `export_loop`) looks like now:

```python
def export(self, list_instance, format, context):
    data = []
    for value in list_instance:
        # Just call `export` on the field; no more inspecting for `export_loop`.
        shaped = self.field.export(value, format, context)
        # Filter out empty values as before...
        ...elif shaped is not None:
            data.append(shaped)
    return data
```

When the process reaches the atomic field, `BaseType.export()` will call either `self.to_primitive()` or `self.to_native()` based on the format.

There are "only" two breaking changes:

* Custom compound types must implement the new interface by turning `to_native` into `convert` and `export_loop` into `export`, but this is very straightforward.

* `Model.to_native()` now actually returns a native value (instance) instead of a dict.

**Comment by bintoro** [comment]
> Something @jvantuyl brought up: is there any real reason why we have the import and export loops instead of recursion? Do they buy us anything?

As I've been refactoring the code, I've found myself wondering the same thing about the standalone functions `validate()`, `to_native()`, and `to_primitive()`.

Since `export_loop` doesn't handle missing keys, any dict you would put through `to_primitive()` is one that came from `to_native()`. But if the only use case is a two-way transformation, why not just instantiate a model?

It's not like it's any more efficient or interoperable to work with a plain dict. All submodels *will* be instantiated in any case, so the only difference is whether the root object remains a dict or not.

**Comment by bintoro** [comment]
Guys, I want to finally release v2.0. Is there anything left on the wishlist that would likely be a backward-incompatible change?

**Comment by lkraider** [comment]
Yes, get rid on Model._initial and split instance raw_data from validated data. My patch only works against 0.9x tho.

**Comment by bintoro** [comment]
Good call on `_initial`. I wonded if anybody's using it these days.

I did some Git archeology, and the attribute first appeared in d6543a1215f2f485d95b2a2dd474283aa56e4902 under the name `initial`. Back then it was accompanied by a `reset()` method that would revert the instance's data back to its original state. But `reset()` no longer exists.

I can see some value in retaining a pointer to the original data, although I've never used this feature myself.

On the other hand, it does cause a little bit of memory overhead, as the raw data can't be garbage collected while `_initial` is holding a reference to it.

But other than that, `_initial` seems harmless and can be useful I suppose. Perhaps it could be made opt-in. I'm pretty sure the vast majority of applications don't use it.

@lkraider The data splitting thing is an additive feature, right? If so, I'd prefer to do it post v2.0 so as not to delay the release any longer than necessary.

**Comment by bintoro** [comment]
What about `LongType`? Why does it exist? The process of unifying short and long integers for all practical purposes was completed with Python 2.4, released in the year 2004.

**Comment by lkraider** [comment]
Re `_initial`: If the user needs the original data, they can always just add a property dynamically or subclass. Don't see a reason to complicate the API for that.

**Comment by bintoro** [comment]
> Re `_initial`: If the user needs the original data, they can always just add a property dynamically or subclass.

Yeah, I think you're totally right.

Another thing that's a big question mark to me is `Model.__optionsclass__`. I guess one could imagine it being mildly useful... if you ignore the fact that it's not even inherited:

```python
class A(Model):
    __optionsclass__ = MyOptions


class B(A):
    class Options:
        ... # These won't use the MyOptions class.
```

If there's enthusiasm to improve on this, we could do a soft deprecation of `__optionsclass__` along with the introduction of a new mechanism that gets looked up on the superclasses as well.

But I wonder if this whole feature makes much sense. If it's only to create static option templates for easy reuse across an application, then that could be accomplished with a more lightweight system:

```python
class MyStandardOptions:
    namespace = 'foo'


class X(Model):
    Options = BaseOptions   # This is already possible.


class Y(Model):
    class Options(MyStandardOptions):  # This would be easy to implement.
        export_order = True
```

**Comment by lkraider** [comment]
I am working on bringing Calculated Types (aka. Serializable `setters`) into current schematics to better support ORMs - namely SchemAlchemy (https://github.com/schematics/schemalchemy/issues/1).

There a few test cases that I am creating that will require some adjustments in schematics. I'll create a pull request with the changes for discussion.
